File: ios/Classes/AppRestriction/RestrictionModeUpsertCache.swift
import Foundation

struct RestrictionCachedMode {
    let modeId: String
    let isEnabled: Bool
    let blockedAppIds: [String]
}

enum RestrictionModeUpsertCache {
    private static var order: [String] = []
    private static var modes: [String: RestrictionCachedMode] = [:]
    private static let maxEntries = 50
    private static let lock = NSLock()

    static func upsert(_ mode: RestrictionCachedMode) {
        lock.lock()
        defer { lock.unlock() }
        modes[mode.modeId] = mode
        order.removeAll { $0 == mode.modeId }
        order.append(mode.modeId)
        while order.count > maxEntries {
            let evicted = order.removeFirst()
            modes.removeValue(forKey: evicted)
        }
    }

    static func get(modeId: String) -> RestrictionCachedMode? {
        lock.lock()
        defer { lock.unlock() }
        guard let mode = modes[modeId] else {
            return nil
        }
        order.removeAll { $0 == modeId }
        order.append(modeId)
        return mode
    }

    static func remove(modeId: String) {
        lock.lock()
        defer { lock.unlock() }
        modes.removeValue(forKey: modeId)
        order.removeAll { $0 == modeId }
    }
}

File: android/src/main/kotlin/com/example/pauza_screen_time/app_restriction/RestrictionModeUpsertCache.kt
package com.example.pauza_screen_time.app_restriction

internal data class RestrictionCachedMode(
    val modeId: String,
    val blockedAppIds: List<String>,
)

internal object RestrictionModeUpsertCache {
    private const val maxEntries = 50
    private val modes = object : LinkedHashMap<String, RestrictionCachedMode>(maxEntries, 0.75f, true) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, RestrictionCachedMode>?): Boolean {
            return size > maxEntries
        }
    }

    @Synchronized
    fun upsert(mode: RestrictionCachedMode) {
        modes[mode.modeId] = mode
    }

    @Synchronized
    fun get(modeId: String): RestrictionCachedMode? {
        return modes[modeId]
    }

    @Synchronized
    fun remove(modeId: String) {
        modes.remove(modeId)
    }
}
